### ISSUE TITLE
Error `is_non_propositional_predicate::found` on macOS with the OCaml binding

### ISSUE THREAD

**Issue by arbipher** [Z3Prover/z3#5578]
One example crashes using the OCaml API on the upstream master branch. It's not the same as the previous issues #5517 (or #2305).

The error is `libc++abi: terminating with uncaught exception of type is_non_propositional_predicate::found`.
The error only occurs on my macOS (Big Sur 11.5.2). The code works fine on my ubuntu 20.04 with z3 at 4.8.11.

The simplest example to trigger to crash is

```ocaml
open Z3

let ctx = Z3.mk_context []
let solver = Solver.mk_solver ctx None

let () = 
  let x = Arithmetic.Integer.mk_const_s ctx "x" in
  let one = Arithmetic.Integer.mk_numeral_i ctx 1 in
  let eq_x_1 = Boolean.mk_eq ctx one x in
  Z3.Solver.add solver [eq_x_1];
  let _ = Z3.Solver.check solver [] in
  ()
```

The build command is `ocamlfind ocamlopt -o main -thread -package z3 -linkpkg main.ml` and run with `./main`.

In my testing, only adding `x = 1` will trigger the bug. Adding any of `x = x` (of ints) `1 = 1` `x = true`(x is a bool) works fine.

**Comment by NikolajBjorner** [comment]
> The error only occurs on my macOS (Big Sur 11.5.2).

The exception is internal. It is always caught and does not have a way to escape (AFAIK). 
You suggest it is platform specific and could therefore be due to conditions outside of z3 (compiler bugs?).

**Comment by arbipher** [comment]
I see. Thanks for your immediate reply.
Let me post it in the OCaml community to see some feedback.

**Comment by NikolajBjorner** [comment]
It may be your C++ compiler

**Comment by NikolajBjorner** [comment]
If you are able to produce a stack (should be possible, but maybe not easy) you can trace it for exception handlers. There should be a handler for the uncaught exception. is_non_propositional_predicate is defined in src/tactic/probe.cpp it is used in a template that catches exceptions.

**Comment by arbipher** [comment]
This situation also occurs on some other members in my group who use macOS.
Though It's possible we all have some problems with the C++ compiler configuration.

**Comment by arbipher** [comment]
> If you are able to produce a stack (should be possible, but maybe not easy) you can trace it for exception handlers. There should be a handler for the uncaught exception. is_non_propositional_predicate is defined in src/tactic/probe.cpp it is used in a template that catches exceptions.

Sure. Let me do it.

**Comment by arbipher** [comment]
I am able to get the stack trace. I will leave the rest debug work tomorrow.

```shell
λ ./main 
0   main                                0x0000000100f0f0f2 _ZN22is_propositional_probeclERK4goal + 306
1   main                                0x0000000100f0e519 _ZN9and_probeclERK4goal + 41
2   main                                0x0000000100f2838b _ZN13cond_tacticalclERK3refI4goalER11sref_bufferIS1_Lj16EE + 75
3   main                                0x0000000100f18d51 _ZN17and_then_tacticalclERK3refI4goalER11sref_bufferIS1_Lj16EE + 561
4   main                                0x0000000100f26577 _ZN14unary_tacticalclERK3refI4goalER11sref_bufferIS1_Lj16EE + 55
5   main                                0x0000000100f2940f _Z4execR6tacticRK3refI4goalER11sref_bufferIS2_Lj16EE + 63
6   main                                0x0000000100f29833 _Z9check_satR6tacticR3refI4goalERS1_I5modelER10labels_vecR7obj_refI3app11ast_managerERSA_IN18dependency_managerINSC_22expr_dependency_configEE10dependencyESC_ERNSt3__112basic_stringIcNSL_11char_traitsIcEENSL_9allocatorIcEEEE + 227
7   main                                0x0000000100ed907f _ZN12_GLOBAL__N_113tactic2solver15check_sat_core2EjPKP4expr + 1631
8   main                                0x0000000100ed689e _ZN12solver_na2as14check_sat_coreEjPKP4expr + 558
9   main                                0x0000000100edd668 _ZN15combined_solver14check_sat_coreEjPKP4expr + 1160
10  main                                0x0000000100efd6fb _ZN6solver9check_satEjPKP4expr + 75
11  main                                0x00000001009626d9 _ZL13_solver_checkP11_Z3_contextP10_Z3_solverjPKP7_Z3_ast + 1321
12  main                                0x0000000100962159 Z3_solver_check + 265
13  main                                0x000000010083df02 n_solver_check + 162
14  main                                0x00000001007c0a65 camlZ3__check_4241 + 149
libc++abi: terminating with uncaught exception of type is_non_propositional_predicate::found
[1]    25740 abort (core dumped)  ./main
```

**Comment by NikolajBjorner** [comment]
The code pointed to is:

https://github.com/Z3Prover/z3/blob/cbe7dd4a48780db2268ec52a60b880af3a21ea48/src/tactic/probe.cpp#L304
```
class is_propositional_probe : public probe {
public:
    result operator()(goal const & g) override {
        return !test<is_non_propositional_predicate>(g);
    }
};
```

It uses a function "test" defined here:

https://github.com/Z3Prover/z3/blob/cbe7dd4a48780db2268ec52a60b880af3a21ea48/src/tactic/goal.h#L190

It contains a block to catch the exception.
Execution resulting from normal compilation, the exception ::found should be caught.
The fact that it escapes suggests that you are compiling with flags that disable exception handling or otherwise there is some issue between what is in the source code vs. compiled code.

**Comment by arbipher** [comment]
Yeah, I found the code point which you also mentioned in the previous reply.
That's the exact place I add the `backtrace` before it's escaping.

Thanks for the new fact. Let me check it in detail.

**Comment by arbipher** [comment]
Aside: should the original code trigger a `non_propositional_predicate` failure here? Is there any stupidity in the example code?

**Comment by NikolajBjorner** [comment]
code looks legitimate.
You can even try to repro outside of ocaml by opening a log before creating z3 state. 

https://github.com/Z3Prover/z3/blob/8a85cfdb128b1426c3e6bb5be20dc987e64a2e37/src/api/ml/z3.mli#L54

**Comment by arbipher** [comment]
I did the OCaml log, but I cannot parse the log output.

```z3
V "4.8.13.0"
R
C 3
= 0x7ff660409998
R
P 0x7ff660409998
S "trace"
S "true"
C 5
R
P 0x7ff660409998
S "well_sorted_check"
S "true"
C 5
R
P 0x7ff660409998
C 7
= 0x7ff660819e08
R
P 0x7ff660409998
C 4
R
P 0x7ff660819e08
U 2
C 361
R
P 0x7ff660819e08
C 371
R
P 0x7ff660819e08
C 441
= 0x7ff66220a8a8
R
P 0x7ff660819e08
C 371
R
P 0x7ff660819e08
P 0x7ff66220a8a8
C 450
R
S "debug"
C 376
R
P 0x7ff660819e08
S "x"
C 30
R
P 0x7ff660819e08
C 371
R
P 0x7ff660819e08
C 33
= 0x7ff66040c3f8
R
P 0x7ff660819e08
C 371
R
P 0x7ff660819e08
P 0x7ff66040c3f8
C 9
R
P 0x7ff660819e08
$ |x|
P 0x7ff66040c3f8
C 51
= 0x7ff662205098
R
P 0x7ff660819e08
C 371
R
P 0x7ff660819e08
P 0x7ff662205098
C 9
R
P 0x7ff660819e08
C 33
= 0x7ff66040c3f8
R
P 0x7ff660819e08
C 371
R
P 0x7ff660819e08
P 0x7ff66040c3f8
C 9
R
P 0x7ff660819e08
I 1
P 0x7ff66040c3f8
C 154
= 0x7ff66220b728
R
P 0x7ff660819e08
C 371
R
P 0x7ff660819e08
P 0x7ff66220b728
C 9
R
P 0x7ff660819e08
P 0x7ff66220b728
P 0x7ff662205098
C 58
= 0x7ff66220b8d8
R
P 0x7ff660819e08
C 371
R
P 0x7ff660819e08
P 0x7ff66220b8d8
C 9
R
I 1
C 360
R
P 0x7ff660819e08
P 0x7ff66220a8a8
P 0x7ff66220b8d8
C 457
R
P 0x7ff660819e08
C 371
R
P 0x7ff660819e08
P 0x7ff66220a8a8
C 468
```

> some issue between what is in the source code vs. compiled code.
I think this is valuable. I am debugging along with this guidance.

**Comment by arbipher** [comment]
The exact exception point is 

https://github.com/Z3Prover/z3/blob/cbe7dd4a48780db2268ec52a60b880af3a21ea48/src/tactic/probe.cpp#L258

where `n` is checked as not a bool.

I add a printing here. `ast_printer.display(tout, n)` shows `n` is `#6` while the beginning part of `z3.log` shows

```
[tool-version] Z3 4.8.13
[mk-app] #1 true
[mk-app] #2 false
[mk-app] #1 true
[mk-app] #2 false
[mk-app] #3 pi
[mk-app] #4 euler
[mk-app] #5 x
[mk-app] #6 Int
[attach-meaning] #6 arith 1
[mk-app] #7 = #6 #5
[mk-app] #8 bv
```

In the example `x` is a int and `x = 1` is a bool. I can only guess is the test here trying checking `x`has to be a `bool`?

**Comment by NikolajBjorner** [comment]
1. Again: The exception should be caught. It is invoked in code that has exception handling.
2. To replay the log, save it as "z3.log" then run "z3 z3.log"

**Comment by arbipher** [comment]
Thanks for your help.
Now I understand your point 1. It should be caught without breaking the program.

Meanwhile, I am thinking about the problem of linking on mac.

Here is some update:

1. It seems macOS has [dropped](https://developer.apple.com/documentation/xcode-release-notes/xcode-10-release-notes?preferredLanguage=occ) `libstdc++`.  That also explains the original message starts with `libc++abi`. I modified the building script `script/mk_util.py` to replace `-lstdlib` to `-stdlib=libc++ -lc++ `.
2. I am checking whether the z3 building result contains linking to `libstd++` (to check 1 is successful).
3. I am checking how to specify in `ocamlfind ocamlopt -o main -thread -package z3 -linkpkg main.ml --verbose` to avoid using `libstd++`.

I may be in a mess here. The whole debugging contains too many unknowns for me. I haven't worked on C++ project for some years (though I did write some c++ at MSR this summer).

**Comment by arbipher** [comment]
Among the three points, I can fix point 3 (and the change on points 1 and 2 should work). The problem remains.

I recall that standard dynamic libraries on mac are with `.dylib` extension. It's also the case for z3.4.8.1 (though there are other linking problems at that time).
However, now we just have `.so` extension of the OCaml libraries on mac. This may suggest someplace in the building script for OCaml binding may ignore the mac. Let me check it.

**Comment by arbipher** [comment]
Finally, I think I got the point: **static link doesn't work on macOS.**

1.
The default (in the official package repository) z3 OCaml binding is built with [`--static-lib`](https://github.com/ocaml/opam-repository/blob/f8dfc4e2b32067697b4602ddb80339665e6bf8c9/packages/z3/z3.4.8.11/opam#L9). I may fix it and make a PR there or ask if some community experts could do that.

While macOS may not have a true static linking (some random [source](https://github.com/lanl/LaGriT/issues/208).

2.
`-lstdc++` is also hard-coded in the building [script](https://github.com/Z3Prover/z3/blob/8a85cfdb128b1426c3e6bb5be20dc987e64a2e37/scripts/mk_util.py#L2011), which is not optimal for mac. Since z3 does respect to mac (`Darwin` system), it's better to add linking to `libc++`.

3.
btw, even fixing 1 by using the dynamic lib, my other [testcases](https://github.com/JHU-PL-Lab/ddse/blob/test-generation-dbmc/test/sudu/test_sudu_z3.ml#L98) still work on ubuntu but not on mac with the latest z3. TLDR; they are different coding to use z3.datatype. I will make separate issues with simple examples when 1 and 2 are solved.

**Comment by NikolajBjorner** [comment]
sniff sniff :'(

**Comment by arbipher** [comment]
I discussed with some friends on how packages are on apt/yun/... .
Usually, packages are distributed via dynamic libraries, and static libraries are used `static` in the names.
If we follow the same idea, using z3 (default for dynamic) and z3-static will be more natural.

**Comment by NikolajBjorner** [comment]
It would be good to make sure all these efforts are to common use and we can close this before the issue grows old.
Since the builds haven't included MacOS + ML bindings until now, I would add a pipeline or include ocaml in one of the Mac pipelines. We have one for python build and one for cmake. The python build should be in scope:

```
- job: "MacOSPython"
  displayName: "MacOS build"
  pool:
    vmImage: "macOS-latest"
  steps:
    - script: python scripts/mk_make.py -d --java --dotnet
    - script: |
        set -e
        cd build
        make -j3
        make -j3 examples
        make -j3 test-z3
        ./cpp_example
        ./c_example
        # java -cp api/java/classes; JavaExample
        cd ..

```
So how about (1) pull requests to get the link flags integrated, (2) either review flags to put into build or another pull request to azure-pipelines.yml or stand alone github action with scenarios.

**Comment by k4rtik** [comment]
Hi @arbipher, thanks for working to resolve this issue. Can you summarize what needs to be done to install Z3 on macOS?

I still get errors/warnings when I follow these steps:
```
python scripts/mk_make.py --ml --prefix=$OPAM_SWITCH_PREFIX
cd build; make -j4
make install
```
here's the warning:
```
[...]
Installed /Users/kartik/.opam/__coq-platform.2021.09.0~8.13/lib/stublibs/dllz3ml.so.owner
ocamlfind: [WARNING] You have installed DLLs but the directory /Users/kartik/.opam/__coq-platform.2021.09.0~8.13/lib/stublibs is not mentioned in ld.conf
Installed /Users/kartik/.opam/__coq-platform.2021.09.0~8.13/lib/Z3/META
Z3 was successfully installed.
```

here's the error:
```
$ ocamlfind ocamlc -o Z3example.byte -thread -package z3 -linkpkg Z3Example.ml
File "Z3Example.ml", line 1:
Error: Error on dynamically loaded library: /Users/kartik/.opam/__coq-platform.2021.09.0~8.13/lib/stublibs/dllz3ml.so: dlopen(/Users/kartik/.opam/__coq-platform.2021.09.0~8.13/lib/stublibs/dllz3ml.so, 0x000A): Library not loaded: /Users/kartik/.opam/__coq-platform.2021.09.0~8.13/lib/stublibs/libz3.dylib
  Referenced from: /Users/kartik/.opam/__coq-platform.2021.09.0~8.13/lib/stublibs/dllz3ml.so
  Reason: tried: '/Users/kartik/.opam/__coq-platform.2021.09.0~8.13/lib/stublibs/libz3.dylib' (no such file), '/usr/local/lib/libz3.dylib' (no such file), '/usr/lib/libz3.dylib' (no such file)
```

When I manually copy `build/libz3.dylib` to `/Users/kartik/.opam/__coq-platform.2021.09.0~8.13/lib/stublibs`, it works but the installer should ideally be doing that, no?

**Comment by arbipher** [comment]
@k4rtik 

> it works but the installer should ideally be doing that, no?

Yes, the new opam files would copy `libz3.dylib` and `libz3.so` (any one of the two) into the `path/to/stublibs`.

However, the new opam files are not in the opam repository yet (sorry to trouble you). I was thinking to wait for z3's next release. z3.4.8.13 would have two packages, `z3` and `z3-static` while in z3.4.8.12 the `z3` uses the static binding by default.

Since the fix of this issue is in the upstream but not any release yet, one possible fix is to make packages `z3.dev` and `z3-static.dev` to the opam repository now to support the develop version of z3.

**Comment by NikolajBjorner** [comment]
so time to roll a new release on Thursday?

**Comment by arbipher** [comment]
I see. I have never thought of that but it's possible that there are users who don't use opam (or dune) in the community.
I could check it later.

p.s. one thing in my mind is Nikolaj's reply above, ideally, the current building of OCaml binding should be ported into the CMake one.

**Comment by arbipher** [comment]
> so time to roll a new release on Thursday?

Sounds good to me and I can pull the opam file things in the Thanksgiving break,

**Comment by k4rtik** [comment]
> I see. I have never thought of that but it's possible that there are users who don't use opam (or dune) in the community. I could check it later.

Well, I just followed the instructions from the README: https://github.com/Z3Prover/z3#building-z3-using-make-and-gccclang

> p.s. one thing in my mind is Nikolaj's reply above, ideally, the current building of OCaml binding should be ported into the CMake one.

Yeah, that will be ideal, but probably involves more work.

**Comment by arbipher** [comment]
I see (again). I will also look at the README when making PR next time.

**Comment by k4rtik** [comment]
Hi @arbipher, I created a PR in `opam-repository` for the recent 4.8.13 release. When you get a chance, can you take a look if the changes look okay? I only changed lines 9, 14 and 15 based on your comment above. And the bottom checksums and source package, of course.

**Comment by arbipher** [comment]
@k4rtik Thanks for making it! The PR is [link](https://github.com/ocaml/opam-repository/pull/20065) for people in this thread.
